### PR TITLE
Multi-Server: Handling PSI variable for qsub daemon

### PR DESCRIPTION
**Pull request by nithinj**
#### Describe Bug or Feature
* Handling PSI variable for qsub

#### Describe Your Change
* qsub will use a named pipe to talk to the background daemon. When qsub wants to submit a job, it will check whether a background daemon is running by checking the pipe belongs to that configuration exists. If the file could not be found, qsub will go ahead and create another daemon and a new pipe for the new configuration.
* There can be more than one daemon based on configurations.
* The name of the pipe should be the unique signature of the background daemon server it is talking to.
* Since pbs_server_instances has been added to conf file we need to account that while generating the name of the pipe.
* Since the name of the pipe is restricted to 108 in Linux, a hash of the configuration string is used as the pipe name, in case the string length exceeds 108.

#### Attach Test and Valgrind Logs/Output
* Filename with small PSI
[root@centos8 test-scripts]# qsub -- /bin/sleep 10
get_comm_filename, fname: /var/tmp/pbs_default_0_____centos8,centos8_2

* File name with large PSI
[root@centos8 test-scripts]# PBS_SERVER_INSTANCES=centos8.localhost12345.localdomain:15001,centos8_2.localhost67890.localdomain:15001 qsub -- /bin/sleep 10
get_comm_filename, fname: /var/tmp/pbs_4d05ffff73ffffffff4058ff280e3b090c3bffff

* Repeating the test to see if it generates the same file name
[root@centos8 test-scripts]# PBS_SERVER_INSTANCES=centos8.localhost12345.localdomain:15001,centos8_2.localhost67890.localdomain:15001 qsub -- /bin/sleep 10
get_comm_filename, fname: /var/tmp/pbs_4d05ffff73ffffffff4058ff280e3b090c3bffff

### CODE REVIEW

**Comment by nithinj** [comment]
Requesting review from @hirenvadalia , @raviv1995 , @ashwathraop

File: src/lib/Libifl/pbs_loadconf.c [hunk 1]
```diff
@@ -66,6 +66,8 @@
 
 char *pbs_conf_env = "PBS_CONF_FILE";
 
+#define LARGE_BUF_LEN 4096
```
Comment (ashwathraop) [inline src/lib/Libifl/pbs_loadconf.c:69]: Is there a common header where you can move? You used it in cmd.h also. You could remove double definitions.

File: src/lib/Libifl/pbs_loadconf.c [hunk 2]
```diff
@@ -1188,6 +1190,107 @@ __pbs_loadconf(int reload)
...
+ */
+char *
+psi_to_str(psi_t *psi, int nsvrs)
+{
+	char buf[LARGE_BUF_LEN];
+	int len = 0;
+	int i;
+
+	if (!psi)
+		return strdup("");
```
Comment (ashwathraop) [inline src/lib/Libifl/pbs_loadconf.c:1215]: Why not return NULL?

File: src/lib/Libifl/pbs_loadconf.c [hunk 3]
```diff
@@ -1188,6 +1190,107 @@ __pbs_loadconf(int reload)
...
+		}
+		fclose(fp);
+	}
+	free(conf_file);
+	
+	if (conf_val_out)
+		return conf_val_out;
+
+	/* Finally, resort to the default. */
+	return strdup("");
```
Comment (ashwathraop) [inline src/lib/Libifl/pbs_loadconf.c:1291]: Why not return NULL?

**Comment by nithinj** [comment]
@ashwathraop ,
I've addressed your comments. Please have a look.

**Review by ashwathraop — APPROVED** [review]

File: src/cmds/qsub_sup.c [hunk 4]
```diff
@@ -1410,32 +1411,50 @@ get_comm_filename(char *fname)
...
-		);
+		count = snprintf(fname, MAXPIPENAME, "%s/pbs_", TMP_DIR);
+		len = snprintf(buf, MAXPIPENAME, "%.16s_%lu_%.8s_%.32s_%.16s_%.5s_%s",
+			 ((server_out == NULL || server_out[0] == 0) ? "default" : server_out),
+			 (unsigned long int) getuid(),
+			 cred_name,
+			 get_conf_path(),
+			 (env_svr == NULL) ? "" : env_svr,
+			 (env_port == NULL) ? "" : env_port,
+			 psi_var ? psi_var : "");
```
Comment (suresh-thelkar) [inline src/cmds/qsub_sup.c:1442]: Just a question. As server and port information is already present in psi do we still need to add evn_svr and evn_port to the pipe name ?

File: src/lib/Libifl/pbs_loadconf.c [hunk 5]
```diff
@@ -1188,6 +1189,107 @@ __pbs_loadconf(int reload)
...
+ * 
+ * @note
+ * return string has to be freed by the caller
+ *
+ * @return char *
+ * @retval !NULL pointer psi string
+ * @retval NULL failure
+ */
+char *
+psi_to_str(psi_t *psi, int nsvrs)
```
Comment (suresh-thelkar) [inline src/lib/Libifl/pbs_loadconf.c:1207]: Do we need nsvrs as a separate argument? We can just call get_num_servers() right.

**Review by raviv1995 — CHANGES_REQUESTED** [review]

File: src/cmds/qsub_sup.c (hunk 4 above)
Comment (nithinj) [reply src/cmds/qsub_sup.c:1442]: pipe name should be unique based on the value in the psi.

**Comment by nithinj** [comment]
@suresh-thelkar & @raviv1995 ,
Thanks for the review. I've addressed your comments. Please have a look.

**Review by raviv1995 — APPROVED** [review]

**Comment by suresh-thelkar** [comment]
@nithinj, thanks for making the changes. It looks good to me.

**Comment by nithinj** [comment]
Thanks for the approval, @ashwathraop , @raviv1995 , @suresh-thelkar 

@openpbs/openpbs-maintainers 
Please have a look at this PR.

**Review by agrawalravi90 — APPROVED** [review]

**Pull request merged**